core/src/main/java/org/apache/calcite/rel/metadata/RelMdTableReferences.java

```diff
@@ -183,6 +185,72 @@ protected RelMdTableReferences() {}
     return result;
   }
 
+  /**
+   * Table references from Intersect.
+   */
+  public Set<RelTableRef> getTableReferences(Intersect intersect, final RelMetadataQuery mq) {
+    final Set<RelTableRef> result = new HashSet<>();
+
+    final Multimap<List<String>, RelTableRef> qualifiedNamesToRefs = HashMultimap.create();
+    for (RelNode input : intersect.getInputs()) {
+      final Map<RelTableRef, RelTableRef> currentTablesMapping = new HashMap<>();
+      final Set<RelTableRef> inputTableRefs = mq.getTableReferences(input);
+      if (inputTableRefs == null) {
+        return null;
+      }
+      for (RelTableRef tableRef : inputTableRefs) {
+        int shift = 0;
+        Collection<RelTableRef> lRefs = qualifiedNamesToRefs.get(
+                tableRef.getQualifiedName());
+        if (lRefs != null) {
+          shift = lRefs.size();
+        }
+        RelTableRef shiftTableRef = RelTableRef.of(
+                tableRef.getTable(), shift + tableRef.getEntityNumber());
+        assert !result.contains(shiftTableRef);
+        result.add(shiftTableRef);
+        currentTablesMapping.put(tableRef, shiftTableRef);
+      }
+      for (RelTableRef newRef : currentTablesMapping.values()) {
+        qualifiedNamesToRefs.put(newRef.getQualifiedName(), newRef);
+      }
+    }
+    return result;
+  }
+
+  /**
+   * Table references from Minus.
+   */
+  public Set<RelTableRef> getTableReferences(Minus minus, final RelMetadataQuery mq) {
+    final Set<RelTableRef> result = new HashSet<>();
+
+    final Multimap<List<String>, RelTableRef> qualifiedNamesToRefs = HashMultimap.create();
+    for (RelNode input : minus.getInputs()) {
+      final Map<RelTableRef, RelTableRef> currentTablesMapping = new HashMap<>();
+      final Set<RelTableRef> inputTableRefs = mq.getTableReferences(input);
+      if (inputTableRefs == null) {
+        return null;
+      }
+      for (RelTableRef tableRef : inputTableRefs) {
+        int shift = 0;
+        Collection<RelTableRef> lRefs = qualifiedNamesToRefs.get(
+                tableRef.getQualifiedName());
+        if (lRefs != null) {
+          shift = lRefs.size();
+        }
+        RelTableRef shiftTableRef = RelTableRef.of(
+                tableRef.getTable(), shift + tableRef.getEntityNumber());
+        assert !result.contains(shiftTableRef);
+        result.add(shiftTableRef);
+        currentTablesMapping.put(tableRef, shiftTableRef);
+      }
+      for (RelTableRef newRef : currentTablesMapping.values()) {
+        qualifiedNamesToRefs.put(newRef.getQualifiedName(), newRef);
+      }
+    }
+    return result;
+  }
+
```

Comment:
The implementation for getTableReferences of `Minus` and `Intersect` is basically the same with `Union`. Can you just put them together?

Reply:
How about just to support SetOp and cover Union and Minus and Intersect.

Reply:
Thanks,You're right.